File: Source/OldCode/Archive/2005 TVA Code Library/Version/CS/TVA.Core/Threading/QueueThread.vb
' James Ritchie Carroll - 2005
Option Explicit On 

Imports System.Threading
Imports System.Reflection

Namespace Threading

    ' This class uses reflection to invoke an existing sub or function on a new thread, its usage
    ' can be as simple as QueueThread.ExecuteMethod(Me, "MyMethod", "param1", "param2", True)
    Public Class QueueThread

        Public ObjectType As Type
        Public Instance As Object
        Public MethodName As String
        Public Parameters As Object()
        Public InvokeAttributes As BindingFlags

        Public Shared Sub ExecuteMethod(ByVal Instance As Object, ByVal MethodName As String, ByVal ParamArray Params As Object())

            Execute(Instance.GetType(), Instance, MethodName, BindingFlags.InvokeMethod Or BindingFlags.Instance Or BindingFlags.Public, Params)

        End Sub

        Public Shared Sub ExecuteNonPublicMethod(ByVal Instance As Object, ByVal MethodName As String, ByVal ParamArray Params As Object())

            Execute(Instance.GetType(), Instance, MethodName, BindingFlags.InvokeMethod Or BindingFlags.Instance Or BindingFlags.NonPublic, Params)

        End Sub

        Public Shared Sub ExecuteSharedMethod(ByVal ObjectType As Type, ByVal MethodName As String, ByVal ParamArray Params As Object())

            Execute(ObjectType, Nothing, MethodName, BindingFlags.InvokeMethod Or BindingFlags.Static Or BindingFlags.Public, Params)

        End Sub

        Public Shared Sub ExecuteNonPublicSharedMethod(ByVal ObjectType As Type, ByVal MethodName As String, ByVal ParamArray Params As Object())

            Execute(ObjectType, Nothing, MethodName, BindingFlags.InvokeMethod Or BindingFlags.Static Or BindingFlags.NonPublic, Params)

        End Sub

        Public Shared Sub ExecutePropertyGet(ByVal Instance As Object, ByVal MethodName As String, ByVal ParamArray Params As Object())

            Execute(Instance.GetType(), Instance, MethodName, BindingFlags.GetProperty Or BindingFlags.Instance Or BindingFlags.Public, Params)

        End Sub

        Public Shared Sub ExecutePropertySet(ByVal Instance As Object, ByVal MethodName As String, ByVal ParamArray Params As Object())

            Execute(Instance.GetType(), Instance, MethodName, BindingFlags.SetProperty Or BindingFlags.Instance Or BindingFlags.Public, Params)

        End Sub

        Public Shared Sub Execute(ByVal ObjectType As Type, ByVal Instance As Object, ByVal MethodName As String, ByVal InvokeAttributes As BindingFlags, ByVal ParamArray Params As Object())

            Dim qt As New QueueThread

            With qt
                .ObjectType = ObjectType
                .Instance = Instance
                .MethodName = MethodName
                .InvokeAttributes = (InvokeAttributes Or BindingFlags.IgnoreCase)
                ReDim .Parameters(Params.Length - 1)
                Params.CopyTo(.Parameters, 0)
            End With

#If ThreadTracking Then
            With ManagedThreadPool.QueueUserWorkItem(AddressOf ThreadProc, qt)
                .Name = "TVA.Threading.QueueThread.ThreadProc()"
            End With
#Else
            ThreadPool.QueueUserWorkItem(AddressOf ThreadProc, qt)
#End If

        End Sub

        Protected Shared Sub ThreadProc(ByVal stateInfo As Object)

            ' Invoke user method
            With DirectCast(stateInfo, QueueThread)
                .ObjectType.InvokeMember(.MethodName, .InvokeAttributes, Nothing, .Instance, .Parameters)
            End With

        End Sub

    End Class

End Namespace

File: Source/OldCode/Archive/2005 TVA Code Library/Version/TVACodeLibrary/TVA.Core_CS/Threading/QueueThread.cs
using System.Diagnostics;
using System.Linq;
using System.Data;
using System.Collections;
using Microsoft.VisualBasic;
using System.Collections.Generic;
using System;
using System.Threading;
using System.Reflection;

// James Ritchie Carroll - 2005


namespace TVA
{
	namespace Threading
	{
		
		// This class uses reflection to invoke an existing sub or function on a new thread, its usage
		// can be as simple as QueueThread.ExecuteMethod(Me, "MyMethod", "param1", "param2", True)
		public class QueueThread
		{
			
			
			public Type ObjectType;
			public object Instance;
			public string MethodName;
			public object[] Parameters;
			public BindingFlags InvokeAttributes;
			
			public static void ExecuteMethod(object Instance, string MethodName, params object[] Params)
			{
				
				Execute(Instance.GetType(), Instance, MethodName, BindingFlags.InvokeMethod || BindingFlags.Instance || BindingFlags.Public, @Params);
				
			}
			
			public static void ExecuteNonPublicMethod(object Instance, string MethodName, params object[] Params)
			{
				
				Execute(Instance.GetType(), Instance, MethodName, BindingFlags.InvokeMethod || BindingFlags.Instance || BindingFlags.NonPublic, @Params);
				
			}
			
			public static void ExecuteSharedMethod(Type ObjectType, string MethodName, params object[] Params)
			{
				
				Execute(ObjectType, null, MethodName, BindingFlags.InvokeMethod || BindingFlags.Static || BindingFlags.Public, @Params);
				
			}
			
			public static void ExecuteNonPublicSharedMethod(Type ObjectType, string MethodName, params object[] Params)
			{
				
				Execute(ObjectType, null, MethodName, BindingFlags.InvokeMethod || BindingFlags.Static || BindingFlags.NonPublic, @Params);
				
			}
			
			public static void ExecutePropertyGet(object Instance, string MethodName, params object[] Params)
			{
				
				Execute(Instance.GetType(), Instance, MethodName, BindingFlags.GetProperty || BindingFlags.Instance || BindingFlags.Public, @Params);
				
			}
			
			public static void ExecutePropertySet(object Instance, string MethodName, params object[] Params)
			{
				
				Execute(Instance.GetType(), Instance, MethodName, BindingFlags.SetProperty || BindingFlags.Instance || BindingFlags.Public, @Params);
				
			}
			
			public static void Execute(Type ObjectType, object Instance, string MethodName, BindingFlags InvokeAttributes, params object[] Params)
			{
				
				QueueThread qt = new QueueThread();
				
				qt.ObjectType = ObjectType;
				qt.Instance = Instance;
				qt.MethodName = MethodName;
				qt.InvokeAttributes = InvokeAttributes || BindingFlags.IgnoreCase;
				qt.Parameters = new object[@Params.Length];
				@Params.CopyTo(qt.Parameters, 0);
				
				#if ThreadTracking
				ManagedThread with_2 = ManagedThreadPool.QueueUserWorkItem(new TVA.Threading.ManagedThreadPool.ParameterizedThreadStart(ThreadProc), qt);
				with_2.Name = "TVA.Threading.QueueThread.ThreadProc()";
				#else
				ThreadPool.QueueUserWorkItem(new System.Threading.WaitCallback(ThreadProc), qt);
				#endif
				
			}
			
			protected static void ThreadProc(object stateInfo)
			{
				
				// Invoke user method
				QueueThread with_1 = ((QueueThread) stateInfo);
				with_1.ObjectType.InvokeMember(with_1.MethodName, with_1.InvokeAttributes, null, with_1.Instance, with_1.Parameters);
				
			}
			
		}
		
	}
}
